crates/astria-sequencer/src/service/consensus.rs

```diff
@@ -189,11 +185,213 @@ impl Consensus {
     }
 
     #[instrument(skip(self))]
-    async fn commit(&mut self) -> Result<response::Commit, BoxError> {
+    async fn commit(&mut self) -> anyhow::Result<response::Commit> {
         let app_hash = self.app.commit(self.storage.clone()).await;
         Ok(response::Commit {
             data: app_hash.0.to_vec().into(),
             ..Default::default()
         })
     }
 }
+
+/// Generates a commitment to the `sequence::Actions` in the block's transactions.
+/// This is required so that a rollup can easily verify that the transactions it
+/// receives are correct (ie. we actually included in a sequencer block, and none
+/// are missing)
+/// It puts this special "commitment" as the first transaction in a block.
+/// When other validators receive the block, they know the first transaction is
+/// supposed to be the commitment, and verifies that is it correct.
+#[instrument]
+fn handle_prepare_proposal(
+    mut prepare_proposal: request::PrepareProposal,
+) -> response::PrepareProposal {
+    let action_commitment = generate_sequence_actions_commitment(&prepare_proposal.txs);
+    let mut txs: Vec<Bytes> = vec![action_commitment.to_vec().into()];
+    txs.append(&mut prepare_proposal.txs);
+    response::PrepareProposal {
+        txs,
+    }
+}
+
+/// Generates a commitment to the `sequence::Actions` in the block's transactions
+/// and ensures it matches the commitment created by the proposer, which
+/// should be the first transaction in the block.
+#[instrument]
+fn handle_process_proposal(process_proposal: &request::ProcessProposal) -> anyhow::Result<()> {
+    let received_action_commitment: [u8; 32] = process_proposal
+        .txs
+        .first()
+        .context("no transaction commitment in proposal")?
+        .to_vec()
+        .try_into()
+        .map_err(|_| anyhow!("transaction commitment must be 32 bytes"))?;
+    let expected_action_commitment =
+        generate_sequence_actions_commitment(&process_proposal.txs[1..]);
+    ensure!(
+        received_action_commitment == expected_action_commitment,
+        "transaction commitment does not match expected",
+    );
+
+    Ok(())
+}
+
+#[cfg(test)]
```

Comment:
I still think we need a snapshot test here. Concern would be that we change the commitment logic accidentally or intentionally, and the output commitment for the same input changes. 

That is either a breaking bug or a major breaking change, right?

Reply:
you mean like with test vectors (ie. these txs result in this commitment?)  yeah I can add this

Reply:
@noot I believe @joroshiba refers to something https://insta.rs/ / https://crates.io/crates/insta

Reply:
I'm fine with just having the commitment inline (depending on size), but yeah I think proper snapshots are better.

Insta looks nice. The nice thing is it's generalizable across other tests and less magic value "where this come from" inputs in the tests, outputs in snapshots. If you see a diff while reviewing it's very obviously a change to think twice about.

Reply:
put a basic test vector unit test for now, but happy to change to `insta` also. does feel a bit overkill for the one test but I see how it's nice, less magic value for sure